### PR TITLE
[Doctrine Bridge] fix UniqueEntityValidator for composite object primary keys

### PR DESCRIPTION
**Pull request by dmaicher**
| Q             | A
| ------------- | ---
| Branch?       | master / 3.1
| Bug fix?      | yes
| New feature?  |no
| BC breaks?    | no
| Deprecations? |no
| Tests pass?   | yes (fail on php 7.1 unrelated?)
| Fixed tickets | https://github.com/symfony/symfony/issues/21274
| License       | MIT
| Doc PR        | -

This PR fixes an issue with the UniqueEntityValidator in case the entity being validated uses a composite primary key via relations to other entities whose classes do not have a `__toString` method.

### CODE REVIEW

File: src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php [hunk 1]
```diff
@@ -138,4 +140,25 @@ public function validate($entity, Constraint $constraint)
...
+     * @param object        $entity
+     *
+     * @return string
+     */
+    private function buildInvalidValueString(ObjectManager $em, ClassMetadata $class, $entity)
+    {
+        $identifiers = array_map(function ($identifier) use ($em) {
+            // identifiers can be objects (without any __toString method) if its a composite PK
+            if (is_object($identifier) && !method_exists($identifier, '__toString')) {
+                return sprintf('(%s)', $this->buildInvalidValueString($em, $em->getClassMetadata(get_class($identifier)), $identifier));
```
Comment (dmaicher) [inline src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php]: is it safe to use the same object manager here for the related class? Not sure if Doctrine supports relations between Classes of different object managers?

File: src/Symfony/Bridge/Doctrine/Tests/Validator/Constraints/UniqueEntityValidatorTest.php [hunk 2]
```diff
@@ -561,4 +563,33 @@ public function testEntityManagerNullObject()
...
+        $entity = new CompositeObjectNoToStringIdEntity($objectOne, $objectTwo);
+
+        $this->em->persist($entity);
+        $this->em->flush();
+
+        $newEntity = new CompositeObjectNoToStringIdEntity($objectOne, $objectTwo);
+
+        $this->validator->validate($newEntity, $constraint);
+
+        $expectedValue = 'Object of class "Symfony\Bridge\Doctrine\Tests\Fixtures\CompositeObjectNoToStringIdEntity" identified by "(Object of class "Symfony\Bridge\Doctrine\Tests\Fixtures\SingleIntIdNoToStringEntity" identified by "1"), (Object of class "Symfony\Bridge\Doctrine\Tests\Fixtures\SingleIntIdNoToStringEntity" identified by "2")"';
```
Comment (dmaicher) [inline src/Symfony/Bridge/Doctrine/Tests/Validator/Constraints/UniqueEntityValidatorTest.php]: should we find a shorter way to build this message? :stuck_out_tongue_closed_eyes:

**Comment by dmaicher** [comment]
@HeahDude as you added this code (https://github.com/symfony/symfony/commit/40c0c52465bd9e8fed49edaad1de954d6547be08) what do you think about this?

**Comment by HeahDude** [comment]
I've discussed this PR with @dmaicher in https://github.com/dmaicher/symfony/issues/1, which is now merged.

This looks good to me.

ping __@symfony/deciders__

**Comment by nicolas-grekas** [comment]
Failure looks related to me (not the messages about getMock, but the one about CompositeObjectNoToStringIdEntity) - maybe a composer.json update needed somewhere?

**Comment by HeahDude** [comment]
@nicolas-grekas I've opened https://github.com/dmaicher/symfony/pull/2 which seems to have fixed the tests, waiting for @dmaicher to merge it.

**Comment by nicolas-grekas** [comment]
@dmaicher can you please squash the history? We can't merge (nor don't want to :) ) merge PR with merges inside.

**Comment by nicolas-grekas** [comment]
(maybe keep just two commits, one for you and one for @HeahDude with rebase -i ?)

**Comment by dmaicher** [comment]
@nicolas-grekas sure :wink: Let's wait for the build and hope its fixed

**Comment by HeahDude** [comment]
Green 🎉

**Comment by dmaicher** [comment]
Done :wink:

**Comment by fabpot** [comment]
I've not checked, but don't we have the same issue in 2.7 and 2.8?

**Comment by HeahDude** [comment]
@fabpot This fix is related to a feature introduced in 3.1 https://github.com/symfony/symfony/pull/15279, and not enough fixed by https://github.com/symfony/symfony/pull/19227 (missing composite identifiers handling, provided here).

**Comment by fabpot** [comment]
Thank you @dmaicher.

**Pull request merged**